crates/linker/src/generate_dylib.rs

```diff
@@ -0,0 +1,421 @@
+use std::error::Error;
+
+use object::elf::{self, SectionHeader64};
+use object::read::elf::{FileHeader, ProgramHeader, SectionHeader, Sym};
+use object::Endianness;
+
+use object::write::elf::Writer;
+use roc_error_macros::internal_error;
+
+// an empty shared library, that we build on top of
+const DUMMY: &[u8] = include_bytes!("../dummy-elf64-x86-64.so");
+
+// index of the dynamic section
+const DYMAMIC_SECTION: usize = 4;
+
+pub fn generate(custom_names: &[String]) -> Result<Vec<u8>, Box<dyn Error>> {
+    let kind = match object::FileKind::parse(DUMMY) {
+        Ok(file) => file,
+        Err(err) => {
+            internal_error!("Failed to parse file: {}", err);
+        }
+    };
+
+    match kind {
+        object::FileKind::Elf64 => copy_file(DUMMY, custom_names),
+        _ => {
+            internal_error!("Not an ELF file");
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Dynamic {
+    tag: u32,
+    // Ignored if `string` is set.
+    val: u64,
+    string: Option<object::write::StringId>,
+}
+
+#[derive(Debug)]
+struct Symbol {
+    in_sym: usize,
+    name: Option<object::write::StringId>,
+}
+
+#[derive(Debug)]
+struct DynamicSymbol {
+    in_sym: usize,
+    name: Option<object::write::StringId>,
+    section: Option<object::write::elf::SectionIndex>,
+    gnu_hash: Option<u32>,
+}
+
+type Section = SectionHeader64<object::Endianness>;
+
+struct Sections {
+    hash: Section,
+    gnu_hash: Section,
+    dynsym: Section,
+    dynstr: Section,
+    dynamic: Section,
+}
+
+#[derive(Default)]
+struct Addresses {
+    hash: u64,
+    gnu_hash: u64,
+    dynsym: u64,
+    dynstr: u64,
+    dynamic: u64,
+}
+
+impl Sections {
+    fn reserve_alloc_sections(
+        &self,
+        writer: &mut Writer,
+        hash_chain_count: u32,
+        gnu_hash_symbol_count: u32,
+        dynamic_count: usize,
+    ) -> ([usize; 5], Addresses) {
+        let sections = self;
+        let endian = Endianness::Little;
+
+        let mut offsets = [0; 5];
+
+        let mut extra_offset: usize = 0;
+        let mut addresses = Addresses::default();
+
+        for (i, out_offset) in offsets.iter_mut().enumerate().take(5) {
```

Comment:
Why is take(5) needed? The array is of length 5 already.

Reply:
that was part of me incrementally rewriting the code to this loop style. Removed now